File: MyAppOne/MyAppOne/Services/MARequestService.m
//
//  MARequestService.m
//  MyAppOne
//
//  Created by Roman Guarino on 10/01/2019.
//  Copyright © 2019 iOS-accelerator. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "MARequestService.h"

@implementation MARequestService

+(void) fetchDataFromUrl:(NSURL *)url success:(void (^)(NSArray *))onSuccess error:(void (^)(NSError *))onError {
    NSURLSession *defaultSession = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSURLSessionDataTask *dataTask = [defaultSession dataTaskWithURL: url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            if(!error) {
                NSArray *dataArray = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
                onSuccess(dataArray);
            } else {
                onError(error);
            }
        }];
        [dataTask resume];
    });
}

@end

File: MyAppOneSwift/MyAppOneSwift/Services/MARequestService.swift
//
//  MARequestService.swift
//  MyAppOneSwift
//
//  Created by Roman Guarino on 25/01/2019.
//  Copyright © 2019 Román Guarino. All rights reserved.
//

import Foundation

class MARequestService {
    
    static func fetchData(url: String, success: @escaping (Data?) -> (), fail: @escaping (Error) -> ()) {
        let defaultSession = URLSession(configuration: .default)
        var dataTask: URLSessionDataTask
        if let urlComponents = URLComponents(string: url) {
            guard let url = urlComponents.url else { return }
            dataTask = defaultSession.dataTask(with: url) { data, response, error in
                DispatchQueue.main.async {
                    if let error = error {
                        fail(error)
                    } else {
                        success(data)
                    }
                }
            }
            dataTask.resume()
        }
    }
}
